### DISCUSSION TITLE
Blog Overview

### DISCUSSION THREAD

**Discussion by giscus** [hbstack/site#729] [General]
# Blog Overview

The blog module provides the basic layouts and templates for blog site.

https://hbstack.dev/modules/blog/overview/

**Comment by gavlinski** [comment]
Hi, can you please add support for `lastmod` manual information or revisioning auto getting from git repo for showing the date of last update of a document?

References:
- https://gohugo.io/methods/page/lastmod/
- https://squidfunk.github.io/mkdocs-material/setup/adding-a-git-repository/?h=author#revisioning

**Reply by razonyang** [reply]
LGTM.

**Reply by razonyang** [reply]
Added simple modules for showing the created and modified dates, I can add more, such as the the git commit message, links for editing and viewing the source file, but it seems dulplicated with the content panel, so I haven't make a decise yet.

- For `blog` layout, please import `github.com/hbstack/blog/modules/revision`.
- For `docs` layout, import `github.com/hbstack/docs/modules/revision` instead.

Tweak the `hb.revision.date_format` to change the date format, currently, it won't rendered in _timeago_ format (sth like * days ago), since extra JS are required.

The i18n files are located at https://github.com/hbstack/revision/tree/main/i18n, I think you might want to translate the items.

**Comment by gavlinski** [comment]
Hi! It's working fine. Thanks!

Here goes some suggestions that shall improve user experience:

1. Show the full text "Revised at..." smaller at the bottom, same size of the date from the top, perhaps no need for icons, no tooltips
  ![image](https://github.com/hbstack/site/assets/1410375/ae2d037d-755b-425a-8edc-d5354bd83b38)
2. Place the last update info together on the top, maybe with a compact date format, tootips for each one as you've already done
  ![image](https://github.com/hbstack/site/assets/1410375/90622809-a2f2-4af3-b206-2a894aed2f7e)

Where can I change/add `hb.revision.date_format` parameter? Can you please provide more details?

**Reply by razonyang** [reply]
> Where can I change/add hb.revision.date_format parameter? Can you please provide more details?

```yaml
// params.yaml
hb:
  revision:
    date_format: ':date_full'
```

> Place the last update info together on the top.

In my opinion, the lastmod wasn't a important thing to show on the top, and put them together seems a little bit weird to me, and was duplicated with the bottom one, but I can create an optional parameter to include/exclude the lastmod.

> Show the full text "Revised at..." smaller at the bottom, same size of the date from the top, perhaps no need for icons, no tooltips

Will make the the icon as optional by adding a new parameter and change the font size, the title (tooltips) will be neccessary/useful for users want to know exact date information (date and time) other than just a date.

**Reply by gavlinski** [reply]
I agree that put them together on the top this way seems a little weird, may confuse the user.

The second approach with a smaller text at the bottom + tooltips with more info looks better.

**Reply by razonyang** [reply]
Available on `github.com/hbstack/revision@v0.2.0`, to remove icons you'll need to set `icons` to `false`.

```yaml
// params.yaml
hb:
  revision:
    # date_format: ':date_full'
    icons: false
    # font_size: "1rem"
```

**Reply by gavlinski** [reply]
Thanks! I have just two more observations:

1. The tooltip text appears to have duplicate time zone information "-0300 -0300":  
  "Revisado em 2023-12-12 14:22:53 -0300 -0300", for `lastmod: 2023-12-12T14:22:53-03:00`
2. Could you please add one more parameter to not show the creation date? Because it already appears at the top.  
  Something like: `show_created: false`

**Reply by razonyang** [reply]
Fixed by `v0.3.0`, it will use localized (`:date_full` and `:time_full`) datetime rather than raw date now.

The `created_at` parameter was added for showing/hiding the date of creation.

**Reply by gavlinski** [reply]
Thank you very much! Now it's working perfectly.

**Reply by razonyang** [reply]
You're welcome.

**Comment by LITUATUI** [comment]
I think I found a bug. 

The blog.list_cols_lg parameter doesn't seem to be working, only the list_cols_md is working.

Example:
```
    list_cols_md: 2
    list_cols_lg: 3
```
I have a large screen (1920x1080) and only 2 article columns. I can change it with the list_cols_md parameter, but the list_cols_lg should work with my screen size.

**Reply by razonyang** [reply]
`list_cols_lg` work only if the sidebar was disabled (by design), but I cant remember the reason.

See https://github.com/hbstack/blog/commit/f5be4a7a16687708c910fbfce79c6b8de315383c#diff-87c53e762995c279bfb9cba355186bc80fdf748d4cd29b3898d06368d7eeb99fR36

**Comment by FuadEfendi** [comment]
I have some issues trying to disable thumbnails, when I use "post_thumbnail: false" then thumbnail id not shown but arrows overlap with link title; using latest "blog". Maybe I can fix it via CSS?

**Reply by razonyang** [reply]
> then thumbnail id not shown but arrows overlap with link title

I didn't get it, could you please provide a screenshot.

**Reply by FuadEfendi** [reply]
I don't know how to upload screenshot... it this class `hb-blog-post-title-link d-block` is center-aligned then it should fix the issue

**Reply by FuadEfendi** [reply]
Hi Razon, you can check https://www.tokenizer.ca/categories/corporate-finance/ for example, on left, between "recent" and "categories"

**Reply by razonyang** [reply]
Are you mean the sidebar's posts module?

![image](https://github.com/hbstack/site/assets/17720932/dea9ec36-bfea-497d-9b16-0d932742ba31)

> I don't know how to upload screenshot...

Yep, Giscus not allow uploading images except using external image services.

**Reply by FuadEfendi** [reply]
yes, and also middle-bottom of this screen, https://www.tokenizer.ca/terms/a/additional-paid-in-capital/

**Reply by razonyang** [reply]
That's indeed a problem, let's continue discussing on hbstack/blog#813.

**Reply by FuadEfendi** [reply]
I am trying to add to "params.yaml" this setting, "meta.date: false", but it doesn't work, I still see dates.

**Reply by FuadEfendi** [reply]
Hi Razon, I am trying to find, https://github.com/hbstack/blog/blob/main/layouts/partials/hb/modules/blog/single.html

When I set "meta: true" of "meta: false" on a page matter, it works; but when I try to set "meta.reading_time: false" it doesn't work. I am also trying to configure it as a site setting (instead of per page)

**Reply by razonyang** [reply]
You can configure those __page params__ via [Cascade ](https://gohugo.io/content-management/front-matter/#cascade), to make pages __extends__ common params.

**Reply by FuadEfendi** [reply]
I have "_index.md" in "content" folder which is my "home page" (it has additional text); when I set it "type: docs" on front matter, it automatically appends all (thousands) of blog posts at the bottom. Which parameter should I use to prevent that? I tried `blog.home.featured_posts = 5` but no change...

**Reply by FuadEfendi** [reply]
Something related to this, https://github.com/hbstack/docs/blob/main/layouts/partials/hb/modules/docs/section.html
It "walks" all documents, so that when I try to have "content/_index.md" as "type: docs" then homepage lists 12,000 (in my specific case) URLs... 5Mb plain HTML.

Making it "index.md" breaks menu items. I am just trying to find way to create homepage. I have tons of documents in non-standard folders which don't have "type" defined

**Reply by FuadEfendi** [reply]
Of course I can override this, in my specific case it is Ok: https://github.com/hbstack/docs/blob/main/layouts/partials/hb/modules/docs/section.html
But I am wondering, maybe there is parameter which can limit this list? "walk-docs-section-tree"

**Reply by FuadEfendi** [reply]
layout: landing?

**Reply by razonyang** [reply]
> maybe there is parameter which can limit this list

There isn't, since the `docs` type is for section and details page.

> layout: landing?

Yeah, the `landing` layout is better for customizing homepage and other pages, you can put anything on pages with shortcodes.

**Comment by gavlinski** [comment]
Hello, I have a few questions:

1. How do I completely disable the entire "Blog section" in a given environment, so that it doesn't display any posts in production, only in development? Is it possible? Or totally disable it in all environments?

2. Is there any practical difference between accessing `/blog` and `/posts`, for example https://hbstack.dev/blog/ and https://hbstack.dev/posts/?

3. My project currently has a folder structure in `/content` in which the subfolder `/content/docs` does not exist, that is, the main content is in subfolders such as `/content/conta`, `/content/ofertas`, and so on. I was able to configure the main menu successfully using `type: docs` property. However, when accessing both the main page and the blog, the same content is being displayed. How can I make it so that, when accessing the blog, only the content of the `/content/blog` and `/content/news` folders is displayed, for example, but not the content of other folders (defined in the `config/_default/params.yaml` file)? I'm almost certain that this behavior used to work exactly like this by default, but it seems to have suddenly stopped.

   ```yaml {title="config/_default/params.yaml"}
   hb:
     blog:
       home: # homepage settings.
         main_sections: # leave it empty to include all regular pages.
           # - docs
           - blog
           - news
           # - conta
           # - ofertas
    ```

**Reply by razonyang** [reply]
> How do I completely disable the entire "Blog section" in a given environment, so that it doesn't display any posts in production, only in development? Is it possible? Or totally disable it in all environments?

Yes, you can achieve this via the `build` option, for example, put the following settings into production configuration, e.g. `config/production/hugo.yaml` or `config/production/config.yaml`.

```yaml
cascade:
  - target:
      path: '{/blog,/blog/**}'
    build:
      list: never
      render: never
```

Read more on https://gohugo.io/content-management/build-options/ and https://gohugo.io/configuration/cascade/

> Is there any practical difference between accessing /blog and /posts, for example https://hbstack.dev/blog/ and https://hbstack.dev/posts/?

There isn't, but I don't remmember why I did that...

> However, when accessing both the main page and the blog, the same content is being displayed...

That's because the homepage and the blog page is using [the same template](https://github.com/hbstack/blog/blob/main/layouts/_default/index.html), I don't think of an ideal solution, or you maybe can use the `landing` layout for homepage. Ideas are welcome.

**Reply by razonyang** [reply]
With the https://github.com/hbstack/blog/releases/tag/v0.41.0, you can specify the `main_sections` parameter other than `blog.home.main_sections` on front matter, e.g. `content/blog/_index.md`

```markdown
---
main_sections:
  - blog
  - news
---
```

**Reply by gavlinski** [reply]
Thank you very much! With this new version and extra config I believe we have achieved the expected result.

The intention was to reproduce the functioning of the pages of the starter theme template:

- https://theme.hbstack.dev/en/ - main page showing **all content** with taxonomies icons activated
   
   ![image](https://github.com/user-attachments/assets/474d3c16-630d-410d-b326-0ed67a54484f)

- https://theme.hbstack.dev/en/blog/ - blog start page showing **only blog posts** (and news) with no taxonomies but with breadcrumbs instead

   ![image](https://github.com/user-attachments/assets/e3ed7173-ad2b-4e10-a7c1-c6c1612b98cf)

How to reactivate the breadcrumbs on the blog's home page?

![image](https://github.com/user-attachments/assets/70f950d1-eec7-4881-b75a-06190624f117)

**Reply by razonyang** [reply]
> How to reactivate the breadcrumbs on the blog's home page?

It's by designed, breadcrumbs doesn't make sence in homepage.

**Reply by gavlinski** [reply]
Another question: when browsing the main page `/` (empty content with default template), the pinned pages are displayed with precedence, always in the first results; however when browsing `/blog` (default location with the same template) the pages are displayed in chronological order disregarding the pinned ones. Both are configured with `featured: false` and `pinned: true`. Is this also by design, missing some parameter or could it be a bug?

**Reply by razonyang** [reply]
By design, the pinned posts (as well as the stats and carousel / featured posts) are shown on [homepage](https://github.com/hbstack/blog/blob/main/layouts/partials/hb/modules/blog/index.html) only, the `/blog` is a [lists page](https://github.com/hbstack/blog/blob/main/layouts/partials/hb/modules/blog/list.html).

You can request a feature on https://github.com/hbstack/blog if you need to show the pinned posts on list pages.

 I think you may misunderstand that the `/blog` is a homepage, which is just a section page, setting `layout` as `index` on `/blog` page may work on your use case.

```yaml
---
layout: index
---
```

But I didn't test, maybe not working.

**Reply by gavlinski** [reply]
Setting the `layout` as `index` on `/blog` didn't work. I'll open an issue requesting this new feature as you suggested.

However, the stats and carousel/featured posts are shown normally on the blog, even though it is a list page as you explained. If the carousel is visible and there is a pinned post, this pinned post also appears next to it, although it does not appear in the paginated list below, which is actually good because it avoids repetitions of the same page.
